src/client/Microsoft.Identity.Client/ManagedIdentity/ManagedIdentityClient.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using Microsoft.Identity.Client.Extensibility;
+using System.Threading.Tasks;
+using System.Threading;
+using System.Linq;
+using Microsoft.Identity.Client.Http;
+using Microsoft.Identity.Client.Internal;
+using Microsoft.IdentityModel.Abstractions;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Utils;
+
+namespace Microsoft.Identity.Client.ManagedIdentity
+{
+    /// <summary>
+    /// Class to initialize a managed identity and identify the service.
+    /// Original source of code: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs
+    /// </summary>
+    internal class ManagedIdentityClient
+    {
+        internal const string MsiUnavailableError =
+            "Authentication with managed identity is unavailable. No managed identity endpoint found.";
+
+        private readonly Lazy<ManagedIdentitySource> _identitySource;
```

Comment:
There is no point in Lazy here, because each request will re-create this logic. 

Now the question is - is this intended? Pretty much like regional where we "cache" the region in a static variable - is it worth caching the MISource in a static variable? 

1. Is it expensive to compute ?
2. Is it liable to change?

Reply:
1. `Is it expensive to compute?` I don't think it will be expensive as it is just trying to get environment variables for different resources. But worst case it will try to create 4 different sources before creating the last one.
2. `Is it liable to change?` The resource where the application is deployed will be the same so I think it makes sense to cache it. If the managed identity is used the source will not change.